Feat: Create order endpoint.

Thiagosavin opened the pull request:
Issue: #9

### What was done
The create order route was included, with all its business rules.

### Why
To allow orders to be created within the system

### How to test
1. Run the project locally.
2. Make a POST request to `/orders` with a valid payload (example in swagger)
3. Ensure the order is saved in the database.

Thiagosavin committed:
feat: Implement order management with create order functionality

Thiagosavin committed:
feat: Enhance order functionality with comprehensive tests and updates to order structure.

dviniciusbonin commented on package.json:
```diff
@@ -27,6 +27,7 @@
     "@prisma/client": "^6.7.0",
     "class-transformer": "^0.5.1",
     "class-validator": "^0.14.2",
+    "fastfood-system": "file:",
```
Para que serve ?

Thiagosavin replied:
Foi colocado por uma extensão. Ja desativei e do retirando.

dviniciusbonin commented on prisma/schema.prisma:
```diff
@@ -37,3 +46,34 @@ model ProductImage {
 }
+
+model Order {
+  id          String         @id @default(uuid())
```
Combinamos de seguir o snake_case para o banco. Para ajustar o nome das colunas e tabelas no prisma, é só utilizar 

```@@map('nome_tabela')``` e @map(`nome_coluna`)

Thiagosavin replied:
Done

dviniciusbonin commented on src/core/order/domain/entities/order.spec.ts:
```diff
@@ -0,0 +1,80 @@
+// order.entity.spec.ts
+import { Order } from './order';
+import { OrderProduct } from './order-product'; // Certifique-se de importar o OrderProduct corretamente
```
Remover comentários de IA

Thiagosavin replied:
Done

dviniciusbonin commented on src/core/order/domain/entities/order.ts:
```diff
@@ -0,0 +1,54 @@
+// order.ts
```
Remover comentário

Thiagosavin replied:
Done

dviniciusbonin commented on src/core/order/domain/entities/order-product.ts:
```diff
@@ -0,0 +1,14 @@
+  productId: string;
+  qtd: number;
+
+  constructor(props: OrderProductProps) {
```
É possível adicionar um produto ao pedido com quantidade 0

Thiagosavin replied:
done

dviniciusbonin commented on src/core/order/domain/entities/order-client.ts:
```diff
@@ -0,0 +1,14 @@
+  readonly clientId?: string;
+
+  constructor(props: OrderClientProps) {
+    this.cpf = props.cpf;
```
CPF pode ser um Value Object

Thiagosavin replied:
Done

Thiagosavin committed:
feat: Refactor order handling and add CPF validation logic

Thiagosavin committed:
feat: Add order-related database migrations and enhance CPF validation handling

Thiagosavin committed:
fix: Update optional chaining for client ID connection in PrismaOrderMapper

Thiagosavin committed:
feat: Enhance PrismaOrderMapper to correctly handle client connection and creation logic

Thiagosavin committed:
feat: Add validation for quantity in OrderProduct constructor

Thiagosavin committed:
feat: Implement order-related database migrations and update PrismaOrderMapper to include client CPF

Thiagosavin committed:
feat: Implement client and product existence checks in repositories and update tests

Thiagosavin committed:
feat: Create orders migration with related tables and foreign key constraints

dviniciusbonin commented on src/core/order/domain/entities/valueobject/cpf.ts:
Faltou teste unitário

Thiagosavin replied:
Nao faltou nao, ele e testado no order. Se rodar o test:cov vai ver que ele ta coberto

dviniciusbonin commented on src/core/order/domain/entities/order-client.ts:
Acho que aqui podemos mudar o nome de OrderClient para Client, para que nossas classes estejam de acordo com a linguagem ubiqua. OrderClient, seria o ator Cliente. Senti falta de nome e email, visto que pode ser cadastrar na hora de fazer o pedido.

Thiagosavin replied:
Ia colocar esse nome mas poderia bater com o nome do client que vainter no dominio client. Quanto aos dados, no pedido combinamos que teriamos apenas cpf, ou clientId, e o email e nome estaria no cadastro do cliente nao?

dviniciusbonin replied:
Não teremos contexto de cliente. O cliente está no contexto de pedidos segundo o event storming

dviniciusbonin commented on src/core/order/application/usecases/create-order.ts:
```diff
@@ -18,6 +18,29 @@ interface CreateOrderOutput {
+        input.client.clientId,
+      );
+      if (!clientExists) {
+        throw new BadRequestException(
```
Aqui na camda de aplicação, podemos ter exceptions customizadas. BadRequest é algo especifico da API

Thiagosavin replied:
Acredito que podemos criar uma issue pra adicionar uma biblioteca melhor nao? Pois adicionar isso aqui iria travar outros

dviniciusbonin replied:
Não precisa de lib.
`throw new Error('Cliente não existe!')`

Thiagosavin committed:
feat: implement order management with client and product validation

- Added Order, Client, and OrderProduct entities with necessary properties and validation.
- Introduced error handling for invalid CPF, empty products, and invalid quantities.
- Created migration for orders, clients, and order_products tables in Prisma.
- Updated create-order use case and related tests to accommodate new structure.
- Implemented exception filter for handling domain-specific errors in HTTP responses.
- Enhanced DTOs for order creation to include product value and quantity.
- Refactored tests to align with new entity structure and validation logic.

Thiagosavin committed:
feat: add order status and client CPF handling in order entity

dviniciusbonin reviewed (approved):
Approved!

Thiagosavin committed:
feat: enhance exception handling in DomainExceptionFilter and remove non-whitelisted validation
